### ISSUE TITLE
Conditional operator null checking does not work as intended in nested objects.

### ISSUE THREAD

**Issue by hanbings** [microsoft/TypeScript#51904]
# Bug Report

### 🔎 Search Terms

Conditional operator, ternary operator, null, null checking

### 🕗 Version & Regression Information

I was unable to test this on prior versions because, I've only recently started learning TypeCcript, I don't understand the TypeScript testing process, and I noticed that [the same issue](https://github.com/microsoft/TypeScript/issues/47851) was reported earlier this year and has yet to be resolved

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAiCmBbA9tA3gKGt6EAuYeAlsNAE7wQAOyAdhPADywA00A6gHwAUWO-eZAGt4tAFy48ZIrQDmLPv2wATQmAmxoAH2i0AriBAKlOAO5k6sie226DIRdACUEgMJhDAIzDAhzNlzojvwUeHpktNDcqgTQALxxdobQAGQp0OaW8Yn6hk7QAPzBJrrwptAAogAewPBUxHT+HDyCIrRstGXQACoAnlTwFWQWZNwA5PDDyGTQMlR6eGNO+WLFJtzcmXLZSSD5RSWH0AD0x9AAFpPwbDFgpQBuk9CCu2tHneUAynrAtVBNXG4rVENzUKzehw+0AAYmAiCBwkxWM0gcIQRkLHInBCcE4ANyOAC+GGJGHgVRoZDws1oeEmADMfPBoO4vD4-MjAph+LduC44LZcg5+FtZHzrIL7I5gbRxZJpNsdELHJMRnKhiNJYZHBAfn8IHLPMhkCB4GBaBgdXr4PBlNx6RJomoNPk4pxoPdkERlPzWSBvL4AZxHIz4fbHaLrK73Z7vb6PP72UHHMBCMBzuGoqrphINdNox6vT63AmAxyAsHSeTKdTQJAYN9fpQIEHZogqKbEKI8DA-WXW9ycMA6PgyD9BKM3lRpPdCMy8BJRzJ5FOZ3PoMoNI58mhoKSeWo+ehyPAwhFnuciBAAHTKPF7xyio+70LhSJC+-EgRo2U7k9nyI8EvG88E-FUplGP9X3PD8H34XUmygZ9-zfZ4x3gMCrSbW1MydAgXXiGMi3jNlA05d1BxMekgWA298UtQ5oMA4CCX4fccFDEBcMjDgC1jYsWVLJNyOPJiLyvTD+FTPB01w7MyFzCC+OIktSPLZpRNPVCgIkh8qwpaZa3AKAYThBEKFbIh207bteyEsiKyCKSRykcdpl4Q5pyIWc6WeRcpGXLUjFXbz11Maxt2PdiVEPKCtJg+wwJFTExTigDxJvUwkpwGVkLEnSQOy7B5Ly+L30SuCcAQ-VSvSxkQAYSSqutHCHSiW4CLdQs41UxMHI0l8yoyprsE47iUqjQjuoEvthMcyilGogrr1Mei3nyliiQYoc0wzNruHkxSRmUnrBLUgcUPPAqwP0mtoDrEzqlqeoiEaESrI7JBbLOvr1K5FMXLHYAJw8kovJ8+d-IVWQguMMG118+AjumSLd2ijdYs09LYK-MwUtq1CcelH8Cau2jQMq4qINJ5ir2vDDKdwa0kLS1D6sah8sNqVrHQ6uATpm+y-ooy7aYgEboDG-aePYAWSN+i6Nt09HpNk-bDsqJSpv4+X+xEhb+CW2j4DWxihuurbCSAA)

### 💻 Code

```ts
class Demo {
    static response<D, W>(
        token: string,
        data: D | null,
        wrong: W | null
    ): Callback<D, W> {
        return (data == null && wrong == null) ?
            new Exception<D, W>(token, new TypeError('error input')) :
            ((wrong == null) ?
                // here, data never to null
                new Success<D, W>(token, data) :
                new Failure<D, W>(token, wrong)
            );
    }
}
```

### 🙁 Actual behavior

`new Success<D, W>(token, data)` In this expression, data should not be null because it was checked in the previous expression.

### 🙂 Expected behavior

I made it work by adding one more expression, but it was obviously superfluous.
```ts
class Demo {
    static response<D, W>(
        token: string,
        data: D | null,
        wrong: W | null
    ): callback.Callback<D, W> {
        return (data == null && wrong == null) ?
            new Exception<D, W>(token, new TypeError('error input')) :
            ((wrong == null) ?
                (data == null) ?
                    new Exception<D, W>(token, new TypeError('error input')) :
                    new Success<D, W>(token, data) :
                new Failure<D, W>(token, wrong)
            );
    }
}
```

**Comment by RyanCavanaugh** [comment]
Our narrowing isn't a counterfactual logic solver and can't handle things like this. FWIW I haven't seen code that couldn't be written more clearly with a single comparison.

**Comment by hanbings** [comment]
> Our narrowing isn't a counterfactual logic solver and can't handle things like this. FWIW I haven't seen code that couldn't be written more clearly with a single comparison.

Ok, I see. Thank you for your answer.
